diff --git a/helix-term/Cargo.toml b/helix-term/Cargo.toml
index b9534c49..71b15332 100644
--- a/helix-term/Cargo.toml
+++ b/helix-term/Cargo.toml
@@ -31,7 +31,7 @@ assets = [
 ]
 
 [features]
-default = ["git"] # Add steel here for development
+default = ["git", "steel"] # Add steel here for development
 unicode-lines = ["helix-core/unicode-lines", "helix-view/unicode-lines"]
 integration = ["helix-event/integration_test"]
 git = ["helix-vcs/git"]
